src/shaders/doubleSphereCamera.frag

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+// ------------ uniform ----------------------
+uniform highp vec2 FocalLength;
+uniform highp vec2 PrincipalPointOffset;
+uniform highp float Alpha;
+uniform highp float Xi;
+
+#if defined(COLOR_TEXTURE)
+uniform samplerCube ColorTexture;
+#endif
+
+#if defined(DEPTH_TEXTURE)
+uniform samplerCube DepthTexture;
+#endif
+
+// ------------ output -----------------------
+#if defined(COLOR_TEXTURE)
+layout(location = OUTPUT_ATTRIBUTE_LOCATION_COLOR) out vec4 fragmentColor;
+#endif
+
+void main(void) {
+  vec3 m;
+  m.xy = (gl_FragCoord.xy - PrincipalPointOffset) / FocalLength;
+  // MUST flip the x axis
+  m.x = -m.x;
+  float r2 = dot(m.xy, m.xy);
+  float sq1 = 1.0 - (2 * Alpha - 1.0) * r2;
+  if (sq1 < 0.0) {
+    discard;
```

Comment:
Question -- how would the shader perform without the `discard`s? As in, GPU time spent. If my knowledge isn't outdated, the `discard`s inhibit certain optimizations in the driver and have a similar effect as if blending would be unnecessarily enabled.

If you're worried about `sqrt()` called with a negative value, that's fine, you just get a NaN, which on most drivers would result in a black color output. Assuming the edges on the output don't affect the consumer, I think even random noise on the edges would be okay.

*Alternatively*, to avoid the need for a discard and to have a control over what's the "outside color", instead of a fullscreen triangle you could render a double circle. This is a common optimization in VR reprojection where the cost of those few extra fragment shader invocations on the edges is high enough that rendering a more complex mesh usually pays off. Not sure here, you'd have to profile, but I don't think it would be worse than with the discards :)

Reply:
Update:
Tried. Yes, visually it will not affect. 
Perf does not not improve. I believe this is because these corner cases are so rare.

Reply:
Good, then it probably doesn't make sense to try the double circle mesh either :)

Reply:
@mosra : I tried it WITHOUT the `discard`. 
>you just get a NaN, which on most drivers would result in a black color output.

No, on my machine, it gives me random colors (sometimes it is light blue, sometimes it is brown etc.) when I resize the window.
However, if I have the "discard" in place, it gives me "black".
With or without the `discard`, the performance is almost the same in this case.

So I put the `discard` back.

**No** `disgard`:
![no-disgard](https://user-images.githubusercontent.com/931093/111848240-d8acba00-88c7-11eb-8398-c5de985e8403.png)

With `disgard`:
![discard](https://user-images.githubusercontent.com/931093/111848268-ebbf8a00-88c7-11eb-8a60-11c1f7b7a53f.png)

Reply:
Are cases with such a huge border around common? Because then rendering just a (double) circle could lead to pretty significant savings. If not (which I assume), then it's not worth caring about I think.

Reply:
>Are cases with such a huge border around common?

@mosra: No, in terms of training, these are rare cases. The most common case is rendered in a square, no border. (see [here](https://user-images.githubusercontent.com/931093/111849867-e5331180-88cb-11eb-9ed4-605eabc9bfd5.png).)